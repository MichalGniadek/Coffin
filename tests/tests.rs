use coffin2::{debug_print, lexer, name_resolution, parser, type_resolution};
use insta::{assert_snapshot, glob};
use std::{fs, path::Path};

fn test_util(path: &Path, spans: bool, variables: bool, types: bool) -> String {
    let code = fs::read_to_string(path).unwrap();
    let lexer = lexer::lex(&code);
    let mut ast = parser::parse(lexer);

    let variables = if variables | types {
        Some(name_resolution::visit(&mut ast))
    } else {
        None
    };

    let types = if types {
        Some(type_resolution::visit(
            &mut ast,
            variables.as_ref().unwrap(),
        ))
    } else {
        None
    };

    // let _module = spirv_generation::visit(&mut ast, &vars, &type_table);

    debug_print::visit(&ast, true, spans, variables.as_ref(), types.as_ref())
}

#[test]
fn parser() {
    glob!(r"insta_parser\*.coff", |path| {
        assert_snapshot!(test_util(path, false, false, false));
    });
}

#[test]
fn parser_spans() {
    glob!(r"insta_parser_spans\*.coff", |path| {
        assert_snapshot!(test_util(path, true, false, false));
    });
}

#[test]
fn variables() {
    glob!(r"insta_variables\*.coff", |path| {
        assert_snapshot!(test_util(path, false, true, false));
    });
}

#[test]
fn types() {
    glob!(r"insta_types\*.coff", |path| {
        assert_snapshot!(test_util(path, false, false, true));
    });
}
