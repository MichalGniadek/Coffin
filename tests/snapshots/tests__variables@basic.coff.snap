---
source: tests/tests.rs
expression: "DebugPrint::visit(&ast, Some(&rodeo), None, Some(&vars), None, Some(errors))"

---
fun fn0@X (arg0@0 : T@X, arg1@1 : T@X)  {
	(let var0@2 = #0)
	(var0@2 = $arg0@0)
	{
		(arg0@0 = #1)
		(let arg0@3 = #2)
		(arg0@3 = #3)
	}
	(arg0@0 = #4)
	(let var1@4 = #5)
	(var1@4 = #6)
	(let var1@5 = #7)
	(var1@5 = #8)
	(arg0@0 = #9)
}
fun fn1@X (arg2@6 : T@X)  {
	(let var1@7 = #0)
	(var1@7 = ($var1@7 + #1))
	(let var1@8 = ($var1@7 + #1))
	(var1@8 = ($var1@8 + #1))
}
fun fn2@X ()  {
	(not_declared@X = #123)
}

===
Undeclared variable. [357..369]
Undeclared type. [14..15]
Undeclared type. [23..24]
Undeclared type. [246..247]
